%\VignetteIndexEntry{methylKit: User Guide}
%\VignetteKeywords{methylBase, methylRaw, calculateDiffMeth}
%\VignettePackage{methylKit}


\documentclass{article}
\title{ methylKit: User Guide}
\usepackage{cite}
\usepackage{hyperref}
\usepackage{url}               % used in bibliography
\bibliographystyle{unsrt}

\usepackage{Sweave}



\begin{document}

\author{Altuna Akalin\\ \texttt{ala2027@med.cornell.edu}\\
\and
Matthias Kormaksson \\ \texttt{mk375@cornell.edu} 
\and Sheng Li  \\ \texttt{shl2018@med.cornell.edu}
}

%%%%%%% YOU may need to introduce result='hide' to compile this again


%%%<<set-options,echo=FALSE,results='hide',cache=FALSE>>=
%%%options(replace.assign=TRUE,width=60)
%%%knit_hooks$set(fig=function(before, options, envir){if (before) par(mar=c(4,4,.1,.1),cex.lab=.95,cex.axis=.9,mgp=c(2,.7,0),tcl=-.3)})
%%%@


\maketitle

\tableofcontents



\section{Introduction}
In this manual, we will show how to use the methylKit package. methylKit is an R package for analysis and annotation of DNA methylation information obtained by high-throughput bisulfite sequencing. The package is designed to deal with sequencing data from RRBS and its variants. But it can potentially handle whole-genome bisulfite sequencing data if proper input format is provided. 

\subsection{DNA methylation}
DNA methylation in vertebrates typically occurs at CpG dinucleotides, however non-CpG Cs are also methylated in certain tissues such as embryonic stem cells. DNA Methylation can act as an epigenetic control mechanism for gene regulation. Methylation can hinder binding of transcription factors and/or methylated bases can be bound by methyl-binding-domain proteins which can recruit chromatin remodeling factors. In both cases, the transcription of the regulated gene will be effected. In addition, aberrant DNA methylation patterns have been associated with many human malignancies and can be used in a predictive manner. In malignant tissues, DNA is either hypo-methylated or hyper-methylated compared to the normal tissue. The location of hyper- and hypo-methylated sites gives a distinct signature to many diseases. Traditionally, hypo-methylation is associated with gene transcription (if it is on a regulatory region such as promoters) and hyper-methylation is associated with gene repression.

\subsection{High-throughput bisulfite sequencing}
Bisulfite sequencing is a technique that can determine DNA methylation patterns. The major difference from regular sequencing experiments is that, in bisulfite sequencing DNA is treated with bisulfite which converts cytosine residues to uracil, but leaves 5-methylcytosine residues unaffected. By sequencing and aligning those converted DNA fragments it is possible to call methylation status of a base. Usually, the methylation status of a base determined by a high-throughput bisulfite sequencing will not be a binary score, but it will be a percentage. The percentage simply determines how many of the bases that are aligning to a given cytosine location in the genome have actual C bases in the reads. Since bisulfite treatment leaves methylated Cs intact, that percentage will give us percent methylation score on that base. The reasons why we will not get a binary response are 1) the probable sequencing errors in high-throughput sequencing experiments 2) incomplete bisulfite conversion 3) (and a more likely scenario) is heterogeneity of samples and heterogeneity of paired chromosomes from the same sample 




\section{Basics}
\subsection{Reading the methylation call files}
We start by reading in the methylation call data from bisulfite
sequencing with \texttt{read} function. Reading in the data this way
will return a methylRawList object which stores methylation
information per sample for each covered base. The methylation call files are basically text
files that contain percent methylation score per base. A typical methylation call file looks like this:
\begin{Schunk}
\begin{Soutput}
        chrBase   chr    base strand coverage freqC  freqT
1 chr21.9764539 chr21 9764539      R       12 25.00  75.00
2 chr21.9764513 chr21 9764513      R       12  0.00 100.00
3 chr21.9820622 chr21 9820622      F       13  0.00 100.00
4 chr21.9837545 chr21 9837545      F       11  0.00 100.00
5 chr21.9849022 chr21 9849022      F      124 72.58  27.42
\end{Soutput}
\end{Schunk}

Most of the time bisulfite sequencing experiments have test and control samples. The test samples can be from a disease tissue while the control samples can be from a healthy tissue. You can read a set of methylation call files that have test/control conditions giving \texttt{treatment} vector option. For sake of subsequent analysis, file.list, sample.id and treatment option should have the same order. In the following example, first two files are have the sample ids "test1" and "test2" and as determined by treatment vector they belong to the same group. The third and fourth files have sample ids "ctrl1" and "ctrl2" and they belong to the same group as indicated by the treatment vector.

\begin{Schunk}
\begin{Sinput}
> library(methylKit)
> file.list=list( system.file("extdata", "test1.myCpG.txt", package = "methylKit"),
+                 system.file("extdata", "test2.myCpG.txt", package = "methylKit"),
+                 system.file("extdata", "control1.myCpG.txt", package = "methylKit"),
+                 system.file("extdata", "control2.myCpG.txt", package = "methylKit") )
> # read the files to a methylRawList object: myobj
> myobj=read(file.list,
+            sample.id=list("test1","test2","ctrl1","ctrl2"),
+            assembly="hg18",
+            treatment=c(1,1,0,0),
+            context="CpG"
+            )
> 
> 
\end{Sinput}
\end{Schunk}

\subsection{Reading the methylation calls from sorted  Bismark alignments}
Alternatively, methylation percentage calls can be calculated from
sorted SAM file(s) from Bismark aligner and read-in to the memory. Bismark is a
popular aligner for bisulfite sequencing reads \cite{Krueger2011}. \texttt{read.bismark} function is designed to read-in Bismark SAM files as \texttt{methylRaw} or \texttt{methylRawList} objects which store per base methylation calls. SAM files must be sorted by chromosome and read position columns, using 'sort' command in unix-like machines will accomplish such a sort easily.

The following command reads a sorted SAM file and creates a \texttt{methylRaw} object for CpG methylation.The user has the option to save the methylation call files to a folder given by \texttt{save.folder} option. The saved files can be read-in using the \texttt{read} function when needed. 

\begin{Schunk}
\begin{Sinput}
> my.methRaw=read.bismark(
+ 	   location=system.file("extdata", "test.fastq_bismark.sorted.min.sam", 
+ 	                    package = "methylKit"),
+              sample.id="test1",assembly="hg18",read.context="CpG",save.folder=getwd())
\end{Sinput}
\end{Schunk}

It is also possible to read multiple SAM files at the same time, check \texttt{read.bismark} documentation.


\subsection{Descriptive statistics on samples}
Since we read the methylation data now, we can check the basic stats about the methylation data such as coverage and percent  methylation. We now have a \texttt{methylRawList} object which contains methylation information per sample. The following command prints out percent methylation statistics for second sample: "test2"

\begin{Schunk}
\begin{Sinput}
> getMethylationStats(myobj[[2]],plot=F,both.strands=F)
\end{Sinput}
\begin{Soutput}
methylation statistics per base
summary:
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   0.00   20.00   82.79   63.17   94.74  100.00 
percentiles:
       0%       10%       20%       30%       40%       50%       60%       70% 
  0.00000   0.00000   0.00000  48.38710  70.00000  82.78556  90.00000  93.33333 
      80%       90%       95%       99%     99.5%     99.9%      100% 
 96.42857 100.00000 100.00000 100.00000 100.00000 100.00000 100.00000 
\end{Soutput}
\end{Schunk}

The following command plots the histogram for percent methylation distribution.The figure below is the histogram and numbers on bars denote what percentage of locations are contained in that bin. Typically, percent methylation histogram should have two peaks on both ends. In any given cell, any given base are either methylated or not. Therefore, looking at many cells should yield a similar pattern where we see lots of locations with high methylation and lots of locations with low methylation.


\begin{center}
%%%<<fig.width=6,fig.height=6.5,out.width='.9\\linewidth'>>=
\begin{Schunk}
\begin{Sinput}
> getMethylationStats(myobj[[2]],plot=T,both.strands=F)
\end{Sinput}
\end{Schunk}
\includegraphics{methylKit-005}
\end{center}



We can also plot the read coverage per base information in a similar way, again numbers on bars denote what percentage of locations are contained in that bin. Experiments that are highly suffering from PCR duplication bias will have a secondary peak towards the right hand side of the histogram.


\begin{center}
%%%<<fig.width=6,fig.height=6.5,out.width='.9\\linewidth'>>=
\begin{Schunk}
\begin{Sinput}
> library ("graphics")
> getCoverageStats(myobj[[2]],plot=T,both.strands=F)
\end{Sinput}
\end{Schunk}
\includegraphics{methylKit-006}
\end{center}

\subsection{Filtering samples based on read coverage}
It might be useful to filter samples based on coverage. Particularly, if our samples are suffering from PCR bias it would be useful to discard bases with very high read coverage. Furthermore, we would also like to discard bases that have low read coverage, a high enough read coverage will increase the power of the statistical tests. The code below filters a \texttt{methylRawList} and discards bases that have coverage below 10X and also discards the bases that have more than 99.9th percentile of coverage in each sample.

\begin{Schunk}
\begin{Sinput}
> filtered.myobj=filterByCoverage(myobj,lo.count=10,lo.perc=NULL,
+                                       hi.count=NULL,hi.perc=99.9)
\end{Sinput}
\end{Schunk}


\section{Comparative analysis}
\subsection{Merging samples}

In order to do further analysis, we will need to get the bases covered in all samples. The following function will merge all samples to one object for base-pair locations that are covered in all samples. Setting \texttt{destrand}=TRUE (the default is FALSE) will merge reads on both strands of a CpG dinucleotide. This provides better coverage, but only advised when looking at CpG methylation (for CpH methylation this will cause wrong results in subsequent analyses). In addition, setting \texttt{destrand}=TRUE will only work when operating on base-pair resolution, otherwise setting this option TRUE will have no effect. The \texttt{unite()} function will return a \texttt{methylBase} object which will be our main object for all comparative analysis. The \texttt{methylBase} object contains methylation information for regions/bases that are covered in all samples.
\begin{Schunk}
\begin{Sinput}
> meth=unite(myobj, destrand=FALSE)
\end{Sinput}
\end{Schunk}

Let us take a look at the data content of methylBase object:
\begin{Schunk}
\begin{Sinput}
> head(meth)
\end{Sinput}
\begin{Soutput}
methylBase object with 6 rows
--------------
              id   chr    start      end strand coverage1 numCs1 numTs1
1 chr21.10011833 chr21 10011833 10011833      +       174    173      1
2 chr21.10011841 chr21 10011841 10011841      +       173    164      9
3 chr21.10011855 chr21 10011855 10011855      +       175    175      0
4 chr21.10011858 chr21 10011858 10011858      +       175    131     44
5 chr21.10011861 chr21 10011861 10011861      +       174    147     27
6 chr21.10011872 chr21 10011872 10011872      +       167    160      7
  coverage2 numCs2 numTs2 coverage3 numCs3 numTs3 coverage4 numCs4 numTs4
1        18     18      0        40     34      6        14     14      0
2        20     19      1        40     18     22        14      8      6
3        21     21      0        39     29     10        14     12      2
4        21     20      1        39     31      8        13      8      5
5        20     15      5        39     13     26        13      9      4
6        20     19      1        39     34      5        14      8      6
--------------
sample.ids: test1 test2 ctrl1 ctrl2 
destranded FALSE 
assembly: hg18 
context: CpG 
treament: 1 1 0 0 
resolution: base 
\end{Soutput}
\end{Schunk}

By default, \texttt{unite} function produces bases/regions covered in all samples. That requirement can be relaxed using "min.per.group" option in \texttt{unite} function.
\begin{Schunk}
\begin{Sinput}
> # creates a methylBase object. Only CpGs covered at least in 1 sample per group will be returned
> # there were two groups defined by the treatment vector given during the creation of myobj treatment=c(1,1,0,0)
> meth.min=unite(myobj,min.per.group=1L)
\end{Sinput}
\end{Schunk}
\subsection{Sample Correlation}
We can check the correlation between samples using \texttt{getCorrelation}. This function will either plot scatter plot and correlation coefficients or just print a correlation matrix

\begin{center}
%%%<<fig.width=6,fig.height=6,out.width='.9\\linewidth'>>=
\begin{Schunk}
\begin{Sinput}
> getCorrelation(meth,plot=T)
\end{Sinput}
\begin{Soutput}
          test1     test2     ctrl1     ctrl2
test1 1.0000000 0.9252530 0.8767865 0.8737509
test2 0.9252530 1.0000000 0.8791864 0.8801669
ctrl1 0.8767865 0.8791864 1.0000000 0.9465369
ctrl2 0.8737509 0.8801669 0.9465369 1.0000000
\end{Soutput}
\end{Schunk}
\includegraphics{methylKit-011}
\end{center}

\subsection{Clustering samples}
We can cluster the samples based on the similarity of their methylation profiles. The following function will cluster the samples and draw a dendrogram.
\begin{center}
%%%<<fig.width=6,fig.height=6,out.width='.9\\linewidth'>>=
\begin{Schunk}
\begin{Sinput}
> clusterSamples(meth, dist="correlation", method="ward", plot=TRUE)
\end{Sinput}
\begin{Soutput}
Call:
hclust(d = d, method = HCLUST.METHODS[hclust.method])

Cluster method   : ward 
Distance         : pearson 
Number of objects: 4 
\end{Soutput}
\end{Schunk}
\includegraphics{methylKit-012}
\end{center}

Setting the plot=FALSE will return a dendrogram object which can be manipulated by users or fed in to other user functions that can work with dendrograms.
\begin{Schunk}
\begin{Sinput}
> hc = clusterSamples(meth, dist="correlation", method="ward", plot=FALSE)
\end{Sinput}
\end{Schunk}
We can also do a PCA analysis on our samples. The following function will plot a scree plot for importance of components.
\begin{center}
%%%<<fig.width=6,fig.height=6,out.width='.9\\linewidth'>>=
\begin{Schunk}
\begin{Sinput}
> PCASamples(meth, screeplot=TRUE)
\end{Sinput}
\end{Schunk}
\includegraphics{methylKit-014}
\end{center}
\ \\ \ \\
\ \\ \ \\
We can also plot PC1 and PC2 axis and a scatter plot of our samples on those axis which will reveal how they cluster.

\begin{center}
%%%%<<fig.width=6,fig.height=6,out.width='.9\\linewidth'>>=
\begin{Schunk}
\begin{Sinput}
> PCASamples(meth)
\end{Sinput}
\end{Schunk}
\includegraphics{methylKit-015}
\end{center}


\subsection{Tiling windows analysis}
For some situations, it might be desirable to summarize methylation information over tiling windows rather than doing base-pair resolution analysis. \texttt{methylKit} provides functionality to do such analysis. The function below tiles the genome with windows 1000bp length and 1000bp step-size and summarizes the methylation information on those tiles. In this case, it returns a \texttt{methylRawList} object which can be fed into \texttt{unite} and \texttt{calculateDiffMeth} functions consecutively to get differentially methylated regions. The tilling function adds up C and T counts from each covered cytosine and returns a total C and T count for each tile.

\begin{Schunk}
\begin{Sinput}
> tiles=tileMethylCounts(myobj,win.size=1000,step.size=1000)
> head(tiles[[1]],3)
\end{Sinput}
\begin{Soutput}
methylRaw object with 3 rows
--------------
                     id   chr   start     end strand coverage numCs numTs
1 chr21.9764001.9765000 chr21 9764001 9765000      *       24     3    21
2 chr21.9820001.9821000 chr21 9820001 9821000      *       13     0    13
3 chr21.9837001.9838000 chr21 9837001 9838000      *       11     0    11
--------------
sample.id: test1 
assembly: hg18 
context: CpG 
resolution: region 
\end{Soutput}
\end{Schunk}

\subsection{Finding differentially methylated bases or regions}
\texttt{calculateDiffMeth()} function is the main function to calculate differential methylation. Depending on the sample size per each set it will either use Fisher's exact or logistic regression to calculate P-values. P-values will be adjusted to Q-values using SLIM method \cite{Wang2011a}.
\begin{Schunk}
\begin{Sinput}
> myDiff=calculateDiffMeth(meth)
\end{Sinput}
\end{Schunk}

After q-value calculation, we can select the differentially methylated regions/bases based on q-value and percent methylation difference cutoffs. Following bit selects the bases that have q-value<0.01 and percent methylation difference larger than 25\%. If you specify \texttt{type="hyper"} or \texttt{type="hypo"} options, you will get hyper-methylated or hypo-methylated regions/bases.
\begin{Schunk}
\begin{Sinput}
> # get hyper methylated bases
> myDiff25p.hyper=get.methylDiff(myDiff,difference=25,qvalue=0.01,type="hyper")
> #
> # get hypo methylated bases
> myDiff25p.hypo=get.methylDiff(myDiff,difference=25,qvalue=0.01,type="hypo")
> #
> #
> # get all differentially methylated bases
> myDiff25p=get.methylDiff(myDiff,difference=25,qvalue=0.01)
\end{Sinput}
\end{Schunk}

We can also visualize the distribution of hypo/hyper-methylated bases/regions per chromosome using the following function. In this case, the example set includes only one chromosome. The \texttt{list} shows percentages of hypo/hyper methylated bases over all the covered bases in a given chromosome.

 
\begin{Schunk}
\begin{Sinput}
> diffMethPerChr(myDiff,plot=FALSE,qvalue.cutoff=0.01, meth.cutoff=25)
\end{Sinput}
\begin{Soutput}
$diffMeth.per.chr
    chr number.of.hypomethylated percentage.of.hypomethylated
1 chr21                       59                     6.126687
  number.of.hypermethylated percentage.of.hypermethylated
1                        75                      7.788162

$diffMeth.all
  percentage.of.hypermethylated number.of.hypermethylated
1                      7.788162                        75
  percentage.of.hypomethylated number.of.hypomethylated
1                     6.126687                       59
\end{Soutput}
\end{Schunk}
 
\subsubsection{Finding differentially methylated bases using multiple-cores}
The differential methylation calculation speed can be increased substantially by utilizing multiple-cores in a machine if available. Both Fisher's Exact test and logistic regression based test are able to use multiple-core option.
\\
The following piece of code will run differential methylation calculation using 2 cores.

\begin{Schunk}
\begin{Sinput}
> myDiff=calculateDiffMeth(meth,num.cores=2)
\end{Sinput}
\end{Schunk}


\section{Annotating differentially methylated bases or regions}
We can annotate our differentially methylated regions/bases based on gene annotation. In this example, we read the gene annotation from a bed file and annotate our differentially methylated regions with that information. This will tell us what percentage of our differentially methylated regions are on promoters/introns/exons/intergenic region. Similar gene annotation can be fetched using \texttt{GenomicFeatures} package available from Bioconductor.org.

\begin{Schunk}
\begin{Sinput}
> gene.obj=read.transcript.features(system.file("extdata", "refseq.hg18.bed.txt", 
+                                            package = "methylKit"))
> #
> # annotate differentially methylated Cs with promoter/exon/intron using annotation data
> #
> annotate.WithGenicParts(myDiff25p,gene.obj)
\end{Sinput}
\begin{Soutput}
summary of target set annotation with genic parts
133 rows in target set
--------------
--------------
percentage of target features overlapping with annotation :
  promoter       exon     intron intergenic 
  27.81955   15.03759   34.58647   57.14286 


percentage of target features overlapping with annotation (with promoter>exon>intron precedence) :
  promoter       exon     intron intergenic 
  27.81955    0.00000   15.03759   57.14286 


percentage of annotation boundaries with feature overlap :
  promoter       exon     intron 
0.28604119 0.02683483 0.17068273 


summary of distances to the nearest TSS :
   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
      5     828   45160   52030   94640  313500 
\end{Soutput}
\end{Schunk}

Similarly, we can read the CpG island annotation and annotate our differentially methylated bases/regions with them.

\begin{Schunk}
\begin{Sinput}
> # read the shores and flanking regions and name the flanks as shores 
> # and CpG islands as CpGi
> cpg.obj=read.feature.flank(system.file("extdata", "cpgi.hg18.bed.txt", 
+                                         package = "methylKit"),
+                            feature.flank.name=c("CpGi","shores"))
> #
> #
> diffCpGann=annotate.WithFeature.Flank(myDiff25p,cpg.obj$CpGi,cpg.obj$shores,
+                                       feature.name="CpGi",flank.name="shores")
\end{Sinput}
\end{Schunk}


\subsection{Regional analysis}
We can also summarize methylation information over a set of defined regions such as promoters or CpG islands. The function below summarizes the methylation information over a given set of promoter regions and outputs a \texttt{methylRaw} or \texttt{methylRawList} object depending on the input.

\begin{Schunk}
\begin{Sinput}
> promoters=regionCounts(myobj,gene.obj$promoters)
> head(promoters[[1]])
\end{Sinput}
\begin{Soutput}
methylRaw object with 6 rows
--------------
                          id   chr    start      end strand coverage numCs
1 chr21.17806094.17808094.NA chr21 17806094 17808094      +     1834     7
2 chr21.10119796.10121796.NA chr21 10119796 10121796      -       79    44
3 chr21.10011791.10013791.NA chr21 10011791 10013791      -     3697  2982
4 chr21.10119808.10121808.NA chr21 10119808 10121808      -       79    44
5 chr21.15357997.15359997.NA chr21 15357997 15359997      -     8613    16
6 chr21.16023366.16025366.NA chr21 16023366 16025366      +     6296     5
  numTs
1  1827
2    35
3   715
4    35
5  8594
6  6291
--------------
sample.id: test1 
assembly: hg18 
context: CpG 
resolution: region 
\end{Soutput}
\end{Schunk}



\subsection{Convenience functions for annotation objects}
After getting the annotation of differentially methylated regions, we can get the distance to TSS and nearest gene name using the \texttt{getAssociationWithTSS} function.

\begin{Schunk}
\begin{Sinput}
> diffAnn=annotate.WithGenicParts(myDiff25p,gene.obj)
> # target.row is the row number in myDiff25p
> head(getAssociationWithTSS(diffAnn))
\end{Sinput}
\begin{Soutput}
     target.row dist.to.feature feature.name feature.strand
60            1             951    NM_199260              -
60.1          2             931    NM_199260              -
60.2          3             838    NM_199260              -
60.3          4             828    NM_199260              -
60.4          5             802    NM_199260              -
60.5          6             723    NM_199260              -
\end{Soutput}
\end{Schunk}

It is also desirable to get percentage/number of differentially methylated regions that overlap with intron/exon/promoters

\begin{Schunk}
\begin{Sinput}
> getTargetAnnotationStats(diffAnn,percentage=TRUE,precedence=TRUE)
\end{Sinput}
\begin{Soutput}
  promoter       exon     intron intergenic 
  27.81955    0.00000   15.03759   57.14286 
\end{Soutput}
\end{Schunk}

We can also plot the percentage of differentially methylated bases overlapping with exon/intron/promoters

\begin{center}
%%%%<<fig.width=6,fig.height=6,out.width='.9\\linewidth'>>=
\begin{Schunk}
\begin{Sinput}
> plotTargetAnnotation(diffAnn,precedence=TRUE,
+     main="differential methylation annotation")
\end{Sinput}
\end{Schunk}
\includegraphics{methylKit-026}
\end{center}

We can also plot the CpG island annotation the same way. The plot below shows what percentage of differentially methylated bases are on CpG islands, CpG island shores and other regions.

\begin{center}
%%%<<fig.width=6,fig.height=6,out.width='.9\\linewidth'>>=
\begin{Schunk}
\begin{Sinput}
> plotTargetAnnotation(diffCpGann,col=c("green","gray","white"),
+        main="differential methylation annotation")
\end{Sinput}
\end{Schunk}
\includegraphics{methylKit-027}
\end{center}

It might be also useful to get percentage of intron/exon/promoters that overlap with differentially methylated bases.

\begin{Schunk}
\begin{Sinput}
> getFeatsWithTargetsStats(diffAnn,percentage=TRUE)
\end{Sinput}
\begin{Soutput}
  promoter       exon     intron 
0.28604119 0.02683483 0.17068273 
\end{Soutput}
\end{Schunk}

\section{methylKit convenience functions}
\subsection{coercion}
Most \texttt{methylKit} objects (methylRaw,methylBase and methylDiff) can be coerced to \texttt{GRanges} objects from \texttt{GenomicRanges} package. Coercing methylKit objects to \texttt{GRanges} will give users additional flexiblity when customising their analyses.

\begin{Schunk}
\begin{Sinput}
> class(meth)
\end{Sinput}
\begin{Soutput}
[1] "methylBase"
attr(,"package")
[1] "methylKit"
\end{Soutput}
\begin{Sinput}
> as(meth,"GRanges")
\end{Sinput}
\begin{Soutput}
GRanges with 963 ranges and 13 metadata columns:
        seqnames               ranges strand   |             id coverage1
           <Rle>            <IRanges>  <Rle>   |       <factor> <integer>
    [1]    chr21 [10011833, 10011833]      +   | chr21.10011833       174
    [2]    chr21 [10011841, 10011841]      +   | chr21.10011841       173
    [3]    chr21 [10011855, 10011855]      +   | chr21.10011855       175
    [4]    chr21 [10011858, 10011858]      +   | chr21.10011858       175
    [5]    chr21 [10011861, 10011861]      +   | chr21.10011861       174
    [6]    chr21 [10011872, 10011872]      +   | chr21.10011872       167
    [7]    chr21 [10011876, 10011876]      +   | chr21.10011876       160
    [8]    chr21 [10011878, 10011878]      +   | chr21.10011878       150
    [9]    chr21 [10011925, 10011925]      -   | chr21.10011925       120
    ...      ...                  ...    ... ...            ...       ...
  [955]    chr21   [9944505, 9944505]      +   |  chr21.9944505        37
  [956]    chr21   [9944663, 9944663]      -   |  chr21.9944663        61
  [957]    chr21   [9959407, 9959407]      +   |  chr21.9959407        44
  [958]    chr21   [9959541, 9959541]      -   |  chr21.9959541        26
  [959]    chr21   [9959569, 9959569]      -   |  chr21.9959569        25
  [960]    chr21   [9959577, 9959577]      -   |  chr21.9959577        25
  [961]    chr21   [9959644, 9959644]      -   |  chr21.9959644        21
  [962]    chr21   [9959650, 9959650]      -   |  chr21.9959650        21
  [963]    chr21   [9967634, 9967634]      -   |  chr21.9967634        10
           numCs1    numTs1 coverage2    numCs2    numTs2 coverage3    numCs3
        <numeric> <numeric> <integer> <numeric> <numeric> <integer> <numeric>
    [1]       173         1        18        18         0        40        34
    [2]       164         9        20        19         1        40        18
    [3]       175         0        21        21         0        39        29
    [4]       131        44        21        20         1        39        31
    [5]       147        27        20        15         5        39        13
    [6]       160         7        20        19         1        39        34
    [7]       148        12        21        18         3        38        24
    [8]       134        16        20        19         1        37        20
    [9]        65        55        37        21        16        68        21
    ...       ...       ...       ...       ...       ...       ...       ...
  [955]         2        35       147        56        91        86        79
  [956]        19        42       116        71        45        45        35
  [957]        17        27       118        58        60        52        49
  [958]        12        14        76        44        32        39        37
  [959]        17         8        77        69         8        40        40
  [960]        25         0        77        71         6        40        40
  [961]         0        21        97        50        47        59        52
  [962]         6        15       103        57        46        59        51
  [963]         0        10        61        25        36        93        62
           numTs3 coverage4    numCs4    numTs4
        <numeric> <integer> <numeric> <numeric>
    [1]         6        14        14         0
    [2]        22        14         8         6
    [3]        10        14        12         2
    [4]         8        13         8         5
    [5]        26        13         9         4
    [6]         5        14         8         6
    [7]        14        11         9         2
    [8]        17        12        12         0
    [9]        47        20         6        14
    ...       ...       ...       ...       ...
  [955]         7        40        25        15
  [956]        10        31        25         6
  [957]         3        40        27        13
  [958]         2        39        32         7
  [959]         0        39        35         4
  [960]         0        39        36         3
  [961]         7        31        14        17
  [962]         8        32        21        11
  [963]        31        56        29        27
  ---
  seqlengths:
   chr21
      NA
\end{Soutput}
\begin{Sinput}
> class(myDiff)
\end{Sinput}
\begin{Soutput}
[1] "methylDiff"
attr(,"package")
[1] "methylKit"
\end{Soutput}
\begin{Sinput}
> as(myDiff,"GRanges")
\end{Sinput}
\begin{Soutput}
GRanges with 963 ranges and 3 metadata columns:
        seqnames               ranges strand   |             id       qvalue
           <Rle>            <IRanges>  <Rle>   |       <factor>    <numeric>
    [1]    chr21 [10011833, 10011833]      +   | chr21.10011833 8.543092e-04
    [2]    chr21 [10011841, 10011841]      +   | chr21.10011841 6.049801e-13
    [3]    chr21 [10011855, 10011855]      +   | chr21.10011855 4.579307e-09
    [4]    chr21 [10011858, 10011858]      +   | chr21.10011858 5.921730e-01
    [5]    chr21 [10011861, 10011861]      +   | chr21.10011861 8.162676e-08
    [6]    chr21 [10011872, 10011872]      +   | chr21.10011872 1.238123e-03
    [7]    chr21 [10011876, 10011876]      +   | chr21.10011876 1.933224e-04
    [8]    chr21 [10011878, 10011878]      +   | chr21.10011878 3.488683e-04
    [9]    chr21 [10011925, 10011925]      -   | chr21.10011925 8.543092e-04
    ...      ...                  ...    ... ...            ...          ...
  [955]    chr21   [9944505, 9944505]      +   |  chr21.9944505 0.000000e+00
  [956]    chr21   [9944663, 9944663]      -   |  chr21.9944663 7.678302e-05
  [957]    chr21   [9959407, 9959407]      +   |  chr21.9959407 4.839266e-08
  [958]    chr21   [9959541, 9959541]      -   |  chr21.9959541 3.145107e-06
  [959]    chr21   [9959569, 9959569]      -   |  chr21.9959569 3.702161e-02
  [960]    chr21   [9959577, 9959577]      -   |  chr21.9959577 4.922906e-01
  [961]    chr21   [9959644, 9959644]      -   |  chr21.9959644 3.291132e-05
  [962]    chr21   [9959650, 9959650]      -   |  chr21.9959650 6.575118e-05
  [963]    chr21   [9967634, 9967634]      -   |  chr21.9967634 1.027764e-03
         meth.diff
         <numeric>
    [1]  10.590278
    [2]  46.670505
    [3]  22.641509
    [4]   2.040816
    [5]  41.197462
    [6]  16.476642
    [7]  24.365768
    [8]  24.693878
    [9]  24.095252
    ...        ...
  [955] -51.017943
  [956] -28.099911
  [957] -36.312399
  [958] -33.559578
  [959] -10.622983
  [960]  -2.084885
  [961] -30.960452
  [962] -28.314428
  [963] -25.862558
  ---
  seqlengths:
   chr21
      NA
\end{Soutput}
\end{Schunk}
\subsection{select}

We can also select rows from \texttt{methylRaw}, \texttt{methylBase} and \texttt{methylDiff} objects with "select" function. An appropriate methylKit object will be returned as a result of "select" function. 
\begin{Schunk}
\begin{Sinput}
> select(meth,1:10) # select first 10 rows of a methylBase object
\end{Sinput}
\begin{Soutput}
methylBase object with 10 rows
--------------
              id   chr    start      end strand coverage1 numCs1 numTs1
1 chr21.10011833 chr21 10011833 10011833      +       174    173      1
2 chr21.10011841 chr21 10011841 10011841      +       173    164      9
3 chr21.10011855 chr21 10011855 10011855      +       175    175      0
4 chr21.10011858 chr21 10011858 10011858      +       175    131     44
5 chr21.10011861 chr21 10011861 10011861      +       174    147     27
6 chr21.10011872 chr21 10011872 10011872      +       167    160      7
  coverage2 numCs2 numTs2 coverage3 numCs3 numTs3 coverage4 numCs4 numTs4
1        18     18      0        40     34      6        14     14      0
2        20     19      1        40     18     22        14      8      6
3        21     21      0        39     29     10        14     12      2
4        21     20      1        39     31      8        13      8      5
5        20     15      5        39     13     26        13      9      4
6        20     19      1        39     34      5        14      8      6
--------------
sample.ids: test1 test2 ctrl1 ctrl2 
destranded FALSE 
assembly: hg18 
context: CpG 
treament: 1 1 0 0 
resolution: base 
\end{Soutput}
\begin{Sinput}
> select(myDiff,20:30) # select rows 10 of a methylDiff object
\end{Sinput}
\begin{Soutput}
methylDiff object with 11 rows
--------------
               id   chr    start      end strand       pvalue       qvalue
20 chr21.10012079 chr21 10012079 10012079      + 1.325366e-07 1.049731e-06
21 chr21.10012089 chr21 10012089 10012089      + 6.797159e-02 1.047612e-01
22 chr21.10012095 chr21 10012095 10012095      + 9.125016e-02 1.324085e-01
23 chr21.10012101 chr21 10012101 10012101      + 8.881784e-16 4.220791e-14
24 chr21.10012696 chr21 10012696 10012696      + 2.253460e-03 6.033165e-03
25 chr21.10012699 chr21 10012699 10012699      + 1.782895e-09 1.955228e-08
   meth.diff
20 26.616915
21  9.564423
22  5.726470
23 39.807824
24  9.684982
25 44.703297
--------------
sample.ids: test1 test2 ctrl1 ctrl2 
destranded FALSE 
assembly: hg18 
context: CpG 
treament: 1 1 0 0 
resolution: base 
\end{Soutput}
\end{Schunk}

\subsection{reorganize}
\texttt{methylBase} and \texttt{methylRawList} can be reorganized by \texttt{reorganize} function. The function can subset the objects based on provided sample ids, it also creates a new treatment vector determining which samples belong to which group. Order of sample ids should match the treatment vector order.
\begin{Schunk}
\begin{Sinput}
> # creates a new methylRawList object
> myobj2=reorganize(myobj,sample.ids=c("test1","ctrl2"),treatment=c(1,0) )
> # creates a new methylBase object
> meth2 =reorganize(meth,sample.ids=c("test1","ctrl2"),treatment=c(1,0) )
\end{Sinput}
\end{Schunk}

\subsection{percMethylation}
Percent methylation values can be extracted from \texttt{methylBase} object by using \texttt{percMethylation} function.
\begin{Schunk}
\begin{Sinput}
> # creates a matrix containing percent methylation values
> perc.meth=percMethylation(meth)
\end{Sinput}
\end{Schunk}

\section{Frequently Asked Questions}
Detailed answers to some of the frequently asked questions and various how-tos 
can be found at \url{http://zvfak.blogspot.com/search/label/methylKit}. 
In addition, \url{http://code.google.com/p/methylkit/} has online documentation
and links to tutorials and other related material. Apart from those here are 
some of the frequently asked questions.

\subsection{How can I select certain regions/bases from \texttt{methylRaw} or \texttt{methylBase} objects ?}
see \texttt{?select} or \texttt{help("[", package = "methylKit")}

\subsection{How can I find if my regions of interest overlap with 
exon/intron/promoter/CpG island etc.?}
Currently, we will be able to tell you if your regions/bases overlap with
the genomic features or not. 
see ?getMembers. 

\subsection{How can I find the nearest TSS associated with my CpGs}
see ?getAssociationWithTSS

\subsection{How do you define promoters and CpG island shores}
Promoters are defined by options at \texttt{read.transcript.features} function. 
The default option is to take -1000,+1000bp around the TSS and you can change that. 
Same goes for CpG islands when reading them in via \texttt{read.feature.flank} function. 
Default is to take 2000bp flanking regions on each side of the CpG island as shores. 
But you can change that as well.

\subsection{What is Bismark SAM output look like, where can I get more info?}
Check tbe Bismark website and there are  also example files that ship with the 
package. Look at their formats and try to run different variations of 
\texttt{read.bismark()} command on example files.

\subsection{How can I reorder or remove samples at/from  \texttt{methylRawList} or \texttt{methylBase} objects ?}
see ?reorganize

\subsection{Should I normalize my data?}
\texttt{methylKit} comes with a simple \texttt{normalizeCoverage()} function to normalize read 
coverage distributions between samples. Ideally, you should first filter bases
with extreme coverage to account for PCR bias using \texttt{filterByCoverage()}
function, then run \texttt{normalizeCoverage()} function to normalize coverages
between samples. These two functions will help reduce the bias in the statistical
tests that might occur due to systematic over-sampling of reads in certain samples.


\subsection{How can I force methylKit to use Fisher's exact test?}
\texttt{methylKit} decides which test to use based on number of samples per group.
In order to use Fisher's exact there must be one sample in each of the test and
control groups. So if you have multiple samples for group, the package will
employ Logistic Regression based test. However, you can use \texttt{pool()} 
function to pool samples in each group so that you have one representative sample
per group. \texttt{pool()} function will sum up number of Cs and Ts in each group.
We recommend using \texttt{filterByCoverage()} and \texttt{normalizeCoverage()} 
functions prior to using \texttt{pool()}.
see ?pool

\subsection{Can use data from other aligners than Bismark in methylKit ?}
Yes, you can. methylKit can read any generic methylation percentage/ratio file
as long as that text file contains columns for chrosome, start, end, strand, 
coverage and number of unmethylated cytosines. However, methylKit can only 
process SAM files from Bismark. For other aligners, you need to get a text file
containing the minimal information described above. Some aligners will come with
scripts or built-in tools to provide such files.
See \url{http://zvfak.blogspot.com/2012/10/how-to-read-bsmap-methylation-ratio.html} 
for how to read methylation ratio files from BSMAP\cite{Xi2009} aligner.


\section{Acknowledgements}
This package is developed at Weill Cornell Medical College by Altuna Akalin 
with important code contributions from Sheng Li and Matthias Kormaksson. 
We wish to thank especially Maria E. Figueroa, Francine Garret-Bakelman, 
Chrisopher Mason and Ari Melnick for their contribution of ideas, data and 
support. Their support and discussions lead to development of methylKit.




\section{R session info}
\begin{Schunk}
\begin{Sinput}
> sessionInfo() 
\end{Sinput}
\begin{Soutput}
R version 2.15.2 (2012-10-26)
Platform: x86_64-apple-darwin9.8.0/x86_64 (64-bit)

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] data.table_1.8.6 methylKit_0.5.6 

loaded via a namespace (and not attached):
[1] BiocGenerics_0.2.0   GenomicRanges_1.10.5 IRanges_1.16.4      
[4] KernSmooth_2.23-8    parallel_2.15.2      stats4_2.15.2       
[7] tools_2.15.2        
\end{Soutput}
\end{Schunk}

\bibliography{Vignette_methylKit.bib}
\end{document}


